src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSeparatedCaseNamingPolicy.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.Json
+{
+    internal abstract class JsonSeparatedCaseNamingPolicy : JsonNamingPolicy
+    {
+        private enum SeparatedCaseState
+        {
+            Start,
+            NewWord,
+            Upper,
+            Lower
+        }
+
+        protected static string ToSeparatedCase(string s, char separator)
+        {
+            if (string.IsNullOrEmpty(s))
+            {
+                return s;
+            }
+
+            StringBuilder sb = new StringBuilder();
```

Comment:
Is this a hot path? If so should it use `ValueStringBuilder`?

Reply:
I copied it as-is from Json.NET as a first step. I guess that naming policy only runs once during class initialization?

Reply:
I wonder what can be used here to reduce the allocation. Unlike camel case, the output length of snake case cannot be predicted.
`ValueStringBuilder`? But we are not in corelib here.
Just maintain a  `stackalloc` buffer here?

Reply:
Previously the team was against `ValueStringBuilder` since it makes libraries to bloat, but this is one of cleanest ways to improve performance. Another one is `ArrayPool<char>`, but it means that you should repeat whole logic of `ValueStringBuilder` which isn't a great idea.

FYI `StringBuilder` has a thread local instance which reused for .NET internals and as I can remember even for string formatting.

Reply:
Worst-case, isn't the required length appx s.Length * 2 (i.e. a separator between every char)?  If so, this can be done simply with:
```C#
char[] arr = ArrayPool<char>.Shared.Rent(s.Length * 2 - 1); // worst-case length for a separator between every char
int pos = 0;
...
arr[pos++] = ...; // every time a char is added
...
string result = new string(arr, 0, pos);
ArrayPool<char>.Shared.Return(arr);
return result;
```